### PR TITLE
[Impeller] order metal samplers according to declared order and not usage order

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/116610

Currently when we generate metal shaders for developer authored fragment shaders, we rely on the declaration order to corespond with the index of each float/sampler. By default samplers seem to get ordered by usage, so we need to add some resource mappings.

### CODE REVIEW

File: impeller/compiler/compiler.cc [hunk 1]
```diff
@@ -35,6 +35,42 @@ static CompilerBackend CreateMSLCompiler(const spirv_cross::ParsedIR& ir,
   sl_options.msl_version =
       spirv_cross::CompilerMSL::Options::make_msl_version(1, 2);
   sl_compiler->set_msl_options(sl_options);
+
+  // Set metal resource mappings to be consistent with location based mapping
+  // used on other backends when creating fragment shaders.
+  if (source_options.remap_samplers) {
```
Comment (chinmaygarde) [inline impeller/compiler/compiler.cc:41]: Can this just be the default?

Comment (jonahwilliams) [reply impeller/compiler/compiler.cc:41]: There are still some compilation errors when using this on the flutter/engine shaders. I need to spend some more time tracking down why that is

Comment (chinmaygarde) [reply impeller/compiler/compiler.cc:41]: SGTM. I'm just catching up on the issue and the conversation in chat. Will take a closer look in the morning.

Comment (jonahwilliams) [reply impeller/compiler/compiler.cc]: I think this works if I also set `enable_decoration_binding` to true. that also fixes the issue with floats being out of order, but I need to verify that our bindings are still correct.

File: impeller/compiler/compiler.cc [hunk 2]
```diff
@@ -20,6 +20,24 @@
 namespace impeller {
 namespace compiler {
 
+static std::optional<spv::ExecutionModel> SourceTypeToExecutionModel(
+    const SourceType& type) {
```
Comment (zanderso) [inline impeller/compiler/compiler.cc:24]: nit: `SourceType` is just an enum, so we'd usually pass it by value.
```suggestion
    SourceType type) {
```

**Comment by jonahwilliams** [comment]
Setting enable_decoration_binding is causing the compute shader headers to generate bogus code. the `ext_res_0` field is ending up with a value of 4294967295 instead of 0, 1, 2.

We may need to map all resource locations and not just samplers

**Comment by jonahwilliams** [comment]
I moved this back to an opt-in flag, there were some confusing issues with compute shaders.

**Comment by jonahwilliams** [comment]
I don't really want to do a rendering test for this. Since this requires a change in the framework to opt in, I could add a test there that scans the output of the generated glowing text shader and verifies the order is as expected. Thoughts?

**Comment by jonahwilliams** [comment]
or I guess I could do that here actually

**Comment by zanderso** [comment]
Could you do something like this https://github.com/flutter/engine/blob/main/testing/dart/fragment_shader_test.dart#L275 ?

**Comment by jonahwilliams** [comment]
No, because it needs to use a metal shader on iOS

File: impeller/fixtures/shader_with_samplers.frag [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// found in the LICENSE file.
+
+// Declare samplers in different order than usage.
+uniform sampler2D textureA;
+uniform sampler2D textureB;
+
+out vec4 frag_color;
+
+void main() {
+  vec4 sample_1 = texture(textureB, vec2(1.0));
```
Comment (jonahwilliams) [inline impeller/fixtures/shader_with_samplers.frag:12]: By default textureB ends up with offset 0 and textureA ends up with offset 1

**Comment by jonahwilliams** [comment]
Failing on CI is a bad sign...

File: testing/dart/fragment_shader_test.dart [hunk 4]
```diff
@@ -332,6 +332,38 @@ void main() async {
     shader.dispose();
   });
 
+  // This test can't rely on actual pixels rendered since it needs to run on a
+  // metal shader on iOS. instead parse the source code.
+  test('impellerc orders samplers in metal shader according to declaration and not usage', () async {
+    final Directory directory = shaderDirectory('iplr-remap');
+    final String data = File(path.join(directory.path, 'shader_with_samplers.frag.iplr')).readAsStringSync();
```
Comment (zanderso) [inline testing/dart/fragment_shader_test.dart:339]: The `.iplr` file is flatbuffer encoded.

Comment (zanderso) [reply testing/dart/fragment_shader_test.dart]: When generating an `.iplr`, the plain text platform specific shader might also be generated nearby. Sorry, I'm blanking on the details, but `impellerc` should be able to generate both at the same time, somehow. Then you can check the bytes of the plain text shader.

**Review by zanderso — APPROVED** [review]
LGTM

**Pull request merged**